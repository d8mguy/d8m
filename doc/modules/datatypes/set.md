## Set

Module set provides an abstract definition of set and an STMap to go's map type.

    [set(T): lst]

where `lst:list(T)`.

    val set = \typegen(T::entity) {
        val setT = extend tuple(...private...) where {
            method $cvt = \() -> list(T) {...}
            method $in = \(elt: T) -> boolean {...}
            method add = \mod(elt: T) -> nothing {...}
            method add = \mod(other: setT) -> nothing {...}
            method remove = \mod(elt: T) -> nothing {...}
            method remove = \mod(elts: list(T)) -> nothing {...}
            method count = \() -> integer {...}
            method isEmpty = \() -> boolean {...}
            method overlaps = \imp(other: setT) -> boolean {...}
            method $= = \(other: setT) -> boolean {...}
            method subset = \(other: setT) -> boolean {...}
            method intersection = \(other: setT) -> setT {...}
            method $+ = \(other: T) -> setT {...}
            method $+ = \(other: setT) -> setT {...}
        }
    }
