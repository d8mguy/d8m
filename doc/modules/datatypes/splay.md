## Splay

The splay module defines type `splaytree`, generic on any ordered type. As usual, this means totally ordered. Splay
trees to not allow for duplicates, so this is not an appropriate implementation of olists, but is good for sets.

In the algorithms literature, splay trees are presented as a type that's efficient in an average sense for accessing
elements, and yet low overhead compared to some other varieties of balanced search trees. Performance measurements to
date don't suggest that splay trees are typically a good choice overall, being slower than most other options for sets,
even when the element type has a total ordering.

This code is based on code in the `Splay_tree` wikipedia article.
Methods `insert` and `delete` return true if they did something.
The `splayOnAccess` attribute controls aspects of when splay operation is called.
Enumeration is done with eachstart.

An STMap is defined for `set(T)`. 

Although splay trees don't seem to be especially performant, this code has some great examples of using pointers to make linked structures, using nilPossible guards, assertions, purification, etc. Thus, it certainly has some value for learning ways of doing certain things in d8m.

    [splaytree(CT): list(CT)]

    val splaytree = \typegen(CT::ordered) {
        method insert = \mod(v: CT) -> boolean {...}
        method delete = \mod(v: CT) -> boolean {...}
        method present = \imp(other: CT) -> boolean {...}
        method count = \mod() -> integer {...}
        method setSplayOnAccess = \mod(flg: boolean) -> nothing {...}
    }
