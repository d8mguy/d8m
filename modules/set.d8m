// Module set provides an abstract definition of set and an STMap to go's map type.

export set

val set = \typegen(T::entity) {
	val setT = extend tuple(private x: list(T)) where {
		assert enumerable(x)
		method $cvt = \() -> list(T) { copy(x) }
		method $in = \(elt: T) { elt in x }
		method add = \mod(elt: T) { unless(elt in x) x.pushb(elt) }
		method add = \mod(other: setT) { each(o^other) self.add(o) }
		method remove = \mod(elt: T) { val inx = x.index(elt); if(inx != nil) x.removeIndex(inx) }
		method remove = \mod(elts: list(T)) { each(e^elts) self.remove(e) }
		method lithook = \mod(c: list(T)) { each(x^c) add(x) }
		method count = \() { x.count }
		method isEmpty = \() { x.count == 0 }
		method overlaps = \imp(other: setT) { var rslt = false; each(e^x) if(e in other) { rslt = true; break }; rslt }
		method $== = \(other: setT) -> boolean { other.count == x.count && all(self.{this in other}) }
		// Note: this is self <= other.
		method subset = \(other: setT) { all(self.{this in other}) }
		// Note: intersection could be more efficient, no need for the uniq-check in "rslt.add(o)". But there are tradeoffs.
		method intersection = \imp(other: setT) { var rslt: setT = []; each(o^other) if(o in x) rslt.add(o); rslt }
		method $+ = \imp(other: T) { var rslt: setT = copy(self); rslt.add(other); rslt }
		method $+ = \imp(other: setT) { var rslt: setT = copy(self); each(o^other) rslt.add(o); rslt }
	}
	setT
}
given(T1::entity, y: T1, aLT: list(T1)) {
	STMap {
		source = set(T1)
		target = Gomap(T1, boolean)
		direct count
		y in self => rvalindexWD(self, y, false)
		self.add(y) => self.lvalindex(y, true)
		self.remove(y) => self.delete(y)
		enumerable(self) => self.keys
	}
}
