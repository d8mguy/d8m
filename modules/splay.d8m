// The splay module defines type `splaytree`, generic on any ordered type. As usual, this means totally ordered. Splay
// trees to not allow for duplicates, so this is not an appropriate implementation of olists, but is good for sets.
//
// In the algorithms literature, splay trees are presented as a type that's efficient in an average sense for accessing
// elements, and yet low overhead compared to some other varieties of balanced search trees. Performance measurements to
// date don't suggest that splay trees are typically a good choice overall, being slower than most other options for sets,
// even when the element type has a total ordering.
//
// This code code in the `Splay_tree` wikipedia article.
// Methods insert, delete return true if they did something.
// The splayOnAccess attribute controls aspects of when splay operation is called.
// Enumeration is done with eachstart.
//
// An STMap is defined for `set(T)`.

//////////////////////////

export splaytree

val splaytree = \typegen(CT::ordered) {
	val STNode: type
	val OSTNode = ortype(nil: nil, main: STNode)
	val STNode = extend tuple(value: CT, left, right, parent: OSTNode) where {
		method next = \() { right == nil ? parent : right }
	}
	extend tuple(private root: OSTNode, private countA: integer, private splayOnAccess: boolean) where {
		private method rotl = \mod(x: STNode) {
			assert rcvrLinked(x)
			var y = x.right
			if(y != nil) {
				var yl = y.left
				x.right = yl
				if(yl != nil) yl.parent = x
				y.parent = x.parent
			}
			if(x.parent == nil) root = y
			else if(x === x.parent.left) x.parent.left = y
			else x.parent.right = y
			if(y != nil) y.left = x
			x.parent = y
		}
		private method rotr = \mod(x: STNode) {
			assert rcvrLinked(x)
			var y = x.left
		    if(y != nil) {
			    x.left = y.right
			    if(y.right != nil) y.right.parent = x
			    y.parent = x.parent
		    }
		    if(x.parent == nil) { root = y }
		    else if(x === x.parent.left) x.parent.left = y
		    else x.parent.right = y
		    if(y != nil) y.right = x
		    x.parent = y
		}
		private method splay = \mod(x: STNode) {
			while(x.parent != nil) {
	    		if(x.parent.parent == nil) if(x.parent.left === x) rotr(x.parent) else rotl(x.parent)
	    		else if(x.parent.left === x && x.parent.parent.left === x.parent) { rotr(x.parent.parent); rotr(x.parent) }
				else if(x.parent.right === x && x.parent.parent.right === x.parent) { rotl(x.parent.parent); rotl(x.parent) }
				else if(x.parent.left === x && x.parent.parent.right === x.parent) { rotr(x.parent); rotl(x.parent) }
				else { rotl(x.parent); rotr(x.parent) }
	    	}
			root = x
		}
		assert purified(splay)

		// replace u with v in the tree
		private method replace = \mod(u: STNode, v: OSTNode) {
            assert rcvrLinked(u)
            assert rcvrLinked(v)
		    if(u.parent == nil) root = v
		    else if(u === u.parent.left) u.parent.left = v
		    else u.parent.right = v
            if(v != nil) v.parent = u.parent
		}

		// search left
		private method subtreeMin = \imp(elt: STNode) {
		    var elt0 = elt
		    while(elt0.left != nil) elt0 = elt0.left
		    elt0
		}

		// return elt containing v or nil
		private method pfind = \imp(v: CT) {
			var elt: OSTNode = root
			while(elt != nil) {
				if(elt.value < v) elt = elt.right
				else if(v < elt.value) elt = elt.left
				else break
			}
			elt
		}
		// insert returns a boolean, true if it inserted
		method insert = \mod(v: CT) {
			if(root == nil) { root = [STNode: v, nil, nil, nil]; countA = 1; return true }
			var p: STNode = root
            loop {
                if(p.value < v && p.right != nil) p = p.right
                else if(v < p.value && p.left != nil) p = p.left
                else break
            }
			if(p.value == v) return false			// v is already present
			countA += 1
			val stn0 = [STNode: v, nil, nil, p]
			if(p.value < v) p.right = stn0 else p.left = stn0
			splay(stn0)
			true
		}
		// delete returns a boolean, true if deleted
		method delete = \mod(v: CT) {
			var elt = pfind(v)
			if(elt == nil) return false
			splay(elt)
			if (elt.left == nil) replace(elt, elt.right)
            else if (elt.right == nil) replace(elt, elt.left)
            else {
                var y = subtreeMin(elt.right)
                if (y.parent !== elt) {
                    replace(y, y.right)
                    y.right = elt.right
                    assert y.right != nil && y.right.parent != nil
                    y.right.parent = y
                }
                replace(elt, y)
                y.left = elt.left
                assert y.left != nil && y.left.parent != nil
                y.left.parent = y
            }
			countA -= 1
			true
		}
		// find an elt == other, true if it's there
		method present = \imp(other: CT) -> boolean {
			if(root == nil) return false
			var elt = root
            loop {
                if(elt.value < other && elt.right != nil) elt = elt.right
                else if(other < elt.value && elt.left != nil) elt = elt.left
                else break
            }
			if(splayOnAccess) splay(elt)
            elt.value == other
        }
		method count = \() { countA }
		method lithook = \mod(ivl: list(CT)) { root = nil; countA = 0; splayOnAccess = false; each(elt^ivl) insert(elt) }
		method setSplayOnAccess = \mod(flg: boolean) { splayOnAccess = flg }
		method eachstart = \imp() {
			val esT: type = extend tuple(cur: OSTNode) where {
				method goleft = \mod() {
					loop { if(cur == nil || cur.left == nil) break; cur = cur.left }
					self
				}
				method eachdone = \() { cur == nil }
				method eachstep = \mod() -> nothing {
					assert cur != nil
					if(cur.right != nil) { cur = cur.right; goleft() }
					else {
						loop {
							val prev = cur
							cur = cur.parent
							if(cur == nil || prev === cur.left) break
						}
					}
				}
				method value = \() { assert cur != nil; cur.value }
			}
			[esT: root].goleft()
		}
	}
}

oncondition defined(set, type) {
    given(OT::ordered, anOT: OT) {
        STMap {
            source = set(OT)
            target = splaytree(OT)
            direct count, lithook
            anOT in self => self.present(anOT)
            self.remove(anOT) => self.delete(anOT)
            self.add(anOT) => self.insert(anOT)
            enumerable(self) => self
        }
    }
}
