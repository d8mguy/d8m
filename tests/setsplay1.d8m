// setsplay1 (aka splay1 mapped from set)
import "splay" melted
import "set" melted
val isetT = set(integer)
var ast = [isetT: 1, 6, 2, 18, 33, 7]
applyST(splaytree(integer), ast)
val trysplay = \mod(st : isetT) {
	var answers : list(boolean) = []
	st.add(15)
	st.add(25)
	answers.pushb(25 in st)	    // true
	st.remove(25)
	answers.pushb(25 in st)   	// false
	st.remove(6)
	answers.pushb(6 in st)		// false
	answers
}
[a~trysplay(ast), b~ast.{this}]
