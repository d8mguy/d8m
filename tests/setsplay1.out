[[true, false, false]; [1, 2, 7, 15, 18, 33]]
